src/lazy/encoder/binary/v1_1/flex_uint.rs

```diff
@@ -46,79 +45,49 @@ impl FlexUInt {
-
+        const COMMON_CASE_INPUT_BYTES_NEEDED: usize = 4;
         // We want to minimize the number of branches that happen in the common case. To do this,
         // we perform a single length check, making sure that the buffer contains enough data to
         // represent a FlexUInt whose continuation bits fit in a single byte (i.e. one with 7 or
-        // fewer bytes of magnitude). If the buffer doesn't have at least 8 bytes in it or the
-        // FlexUInt we find requires more than 8 bytes to represent, we'll fall back to the general
+        // fewer bytes of magnitude). If the buffer doesn't have at least 4 bytes in it or the
+        // FlexUInt we find requires more than 4 bytes to represent, we'll fall back to the general
         // case.
-        if input.len() < COMMON_CASE_INPUT_BYTES_NEEDED || input[0] == 0 {
-            // Calling `read_flex_primitive_as_uint_no_inline` keeps this method small enough that
-            // the code for the common case can be inlined.
-            return Self::read_flex_primitive_as_uint_no_inline(
-                input,
-                offset,
-                "reading a FlexUInt",
-                false,
-            );
+        if input.len() >= COMMON_CASE_INPUT_BYTES_NEEDED {
+            'common_case: {
+                let num_encoded_bytes = input[0].trailing_zeros() as usize + 1;
+                // By branching on particular values, we make the value of `num_encoded_bytes` in their
+                // corresponding arm `const`. This allows us to use `read_n_bytes` to optimize for those
+                // sizes.
+                let mut buffer = [0u8; size_of::<u64>()];
+                match num_encoded_bytes {
+                    1 => Self::read_n_bytes::<1>(input, &mut buffer),
+                    2 => Self::read_n_bytes::<2>(input, &mut buffer),
+                    3 => Self::read_n_bytes::<3>(input, &mut buffer),
+                    4 => Self::read_n_bytes::<4>(input, &mut buffer),
+                    // If the number of encoded bytes isn't 1-4, fall back to the general-purpose
+                    // reading logic.
+                    _ => break 'common_case,
+                };
+                let value = u64::from_le_bytes(buffer).wrapping_shr(num_encoded_bytes as u32);
+                let flex_uint = FlexUInt::new(num_encoded_bytes, value);
+                return Ok(flex_uint);
+            }
         }
-        let flex_uint = Self::read_small_flex_uint(input);
-        Ok(flex_uint)
+        // General-purpose FlexUInt reading logic. Checks for empty input and supports FlexUInts
+        // up to U64::MAX.
+        Self::read_flex_primitive_as_uint(input, offset, "reading a FlexUInt")
     }
 
-    /// Helper method that reads a [`FlexUInt`] with 7 or fewer bytes of magnitude from the buffer.
-    // Caller must confirm that `bytes` has at least 8 bytes.
     #[inline]
-    fn read_small_flex_uint(bytes: &[u8]) -> FlexUInt {
```

Comment:
🪧 This (removed) method was the decoder for `FlexUInt`s up to 7 bytes in size. If the magnitude was only 7 bytes, then the complete serialized form of the `FlexUInt` could fit in a `u64`, simplifying decoding.